### DISCUSSION TITLE
NFC: Reading/writing on iOS unreliable

### DISCUSSION THREAD

**Discussion by ninaDeimos** [capawesome-team/capacitor-plugins#192] [Q&A]
Hi,

We are currently trying to develop a proof-of-concept app for reading and writing NFC tags (ISO 15693) on iOS and Android. On Android, everything works fine, but we are experiencing a lot of issues with the plugin on iOS. On iOS, reading as well as writing the memory of an ISO 15693 tag with the `transceive` function only works about 10 to 15 percent of times. The same function, listed below, is called for each read attempt. The following errors keep occuring, ordered in the frequency of occurances:
- Tag is not connected:
```
{"errorMessage":"Tag is not connected","message":"Tag is not connected"}
```
- Stack Error:

```
[NFXTagReaderSession tranceive:tagUpdate:error:]:893 Error Domain=NFCError Code=100 "Stack error" UserInfo={NSLocalizedDescription=Stack Error, NSUnderlyingError=0x3002cf120 {Error Domain=nfcd Code=54 "Session not active" UserInfo={NSLocalizedDescription=Session not active}}}
```

Our testing environment consists of the following components:
- iPhone XR A2105 with iOS 17.5.21F79
- M24LR04E-R NFC type 5 ISO/IEC 15693 tag from STMicroelectronics with 20 bytes used, no CC-File and zero NDEF records
- Our Ionic 7.2.0 app with TypeScript 5.2.2, Angular 17.0.2 and capacitor-nfc 6.1.0:
```typescript

export function bytesToHex(bytes: number[]): string{
  let hex: string = '';
  let current: string = '';
  for(let i=0; i<bytes.length; i++){
    current = bytes[i].toString(16).toUpperCase();
    while(current.length < 2){
      current = '0' + current;
    }
    hex += current;
  }
  return hex;
}


export function nfcRead(): Promise<Result>{
  return new Promise((resolve, reject) => {
    try {
      let platform: string = Capacitor.getPlatform();
      if(platform === 'web'){
        setTimeout(() => resolve(new Result('10 33 00 FE  00 00 00 0C  58 52 68 74  47 67 00 00  4D 52 00 00'.replaceAll(' ', ''))), 3000);
      } else {
        Nfc.addListener('nfcTagScanned', async (event) => {
          if(platform === 'android'){
            Nfc.connect({techType: NfcTagTechType.NfcV}).then(
              () => {
                Nfc.transceive({
                  data: [0x02, 0x23, 0x00, 0x04]
                }).then(
                  (value) => {
                    console.log(value);
                    try {
                      Nfc.close();
                      Nfc.stopScanSession();
                    } catch (error) {
                      console.log(error);
                    }
                    resolve(new Result(bytesToHex(value.response).substring(2)));
                  }
                ).catch(
                  (reason) => {
                    console.log(reason);
                    try {
                      Nfc.close();
                      Nfc.stopScanSession();
                    } catch (error) {
                      console.log(error);
                    }
                    reject(reason);
                  }
                );
              }
            ).catch(
              (reason) => {
                console.log(reason);
                try {
                  Nfc.close();
                  Nfc.stopScanSession();
                } catch (error) {
                  console.log(error);
                }
                reject(reason);
              }
            );
          } else if(platform === 'ios'){
            Nfc.transceive({
              techType: NfcTagTechType.NfcV,
              data: [0x00, 0x05],
              iso15693CommandCode: 0x23,
              iso15693RequestFlags: [Iso15693RequestFlag.highDataRate],
            }).then(
              (value) => {
                console.log(value);
                try {
                  Nfc.stopScanSession();
                } catch (error) {
                  console.log(error)
                }
                resolve(new Result(bytesToHex(value.response).substring(0, 40)));
              }
            ).catch(
              (reason) => {
                console.log(reason);
                try {
                  Nfc.stopScanSession();
                } catch (error) {
                  console.log(error)
                }
                reject(reason);
              }
            );
          } else {
            try {
              Nfc.stopScanSession();
            } catch (error) {
              console.log(error)
            }
            reject('Unknown plaform `' + platform + '`')
          }
          console.log(event.nfcTag)
        });
        Nfc.startScanSession();
      }
    } catch (error) {
      console.log(error);
      reject(error);
    }
  })
}
 
```
The difference between the Android [0x02, 0x23, 0x00, 0x04] and iOS [0x23], [0x00, 0x05] is intentional, although we don't know why this difference is required to get the same result.

Has anyone managed to reliably read and write ISO 15693 tags on iOS?

Any help is greatly appreciated. Thanks!

**Comment by gion-andri** [comment]
I thoroughly tested the plugin today and encountered the same issues on my end. I attempted to execute transceive commands on ISO15693-based tags using NFC cards from Beointech and the M24LR04 from ST.

On Android, everything works fine. However, on iOS, retrieving any data is nearly impossible. After at least 100 tests, I only received a valid result once.

From extensive debugging, I suspect the issue is related to timing:

- The plugin attempts to read NDEF data for every tag by default, which could be causing delays. It would be beneficial to have an option to disable this to save time.
- After attempting to read the NDEF data and finding none, the plugin returns to the app, which then issues further instructions. By the time these instructions are executed, the connection to the card is already lost.

To validate my assumption, I modified the `tagReaderSession` delegate function to include the `iso15693Tag.readMultipleBlocks` command directly, replacing the NDEF reading logic. With this modification, reading worked perfectly every time.

It seems necessary to allow the plugin to be configured on how to handle found NFC tags before it starts scanning.

Additionally, the read single block command (0x20) and the write single block command (0x21) are not implemented.

@capawesome-team, do you have any recommendations on how to address these issues?

**Reply by robingenz** [reply]
Please send us an email to [support@capawesome.io](mailto:support@capawesome.io) with your changes. I will take a look.

**Comment by dcwang-lab** [comment]
解决了吗，我也遇到过这种问题，可以发我邮箱2099961945@qq.com 看是否能帮您。
